src/event/event_windows.c

```diff
@@ -830,20 +799,142 @@ _dispatch_event_loop_timer_delete(dispatch_timer_heap_t dth DISPATCH_UNUSED,
+	uint64_t minDelayWithLeeway = ~(uint64_t)0;
+	BOOL didFireTimer = FALSE;
+
+#if DEBUG_TIMERS
+	printf("[%lx] Runnimg timers\n", GetCurrentThreadId());
+#endif
+
+	// Fire any timer events that have passed, and work out the
+	// minimum delay until the next one we need to deal with, taking
+	// leeway into account (e.g. a timer that needs to fire in 10ms
+	// with 0ms leeway will take priority over a timer that needs to
+	// fire in 1ms with 100ms leeway).
+	for (int clock = 0; clock < DISPATCH_CLOCK_COUNT; ++clock) {
+#if DEBUG_TIMERS
+		printf("Clock %d: ", clock);
+#endif
+
+		if (!_dispatch_windows_timeout[clock].bArmed) {
+#if DEBUG_TIMERS
+			printf("not armed\n");
+#endif
+			continue;
+		}
+
+		uint64_t now = _dispatch_time_now_cached(clock, nows);
+#if DEBUG_TIMERS
+		printf("current time %"PRIu64", ", now);
+#endif
+		if (_dispatch_windows_timeout[clock].fireTime <= now) {
+#if DEBUG_TIMERS
+			uint64_t lateness = now - _dispatch_windows_timeout[clock].fireTime;
+			printf("firing timer (late by %"PRIu64")\n", lateness);
+#endif
+			didFireTimer = TRUE;
+			_dispatch_event_merge_timer(clock);
+			continue;
+		}
+
+		uint64_t delay = _dispatch_windows_timeout[clock].fireTime - now;
+		uint64_t leeway = _dispatch_windows_timeout[clock].leeway;
+		uint64_t delayWithLeeway;
+
+#if DEBUG_TIMERS
+		printf("delay %"PRIu64", leeway %"PRIu64"\n", delay, leeway);
+#endif
+
+		// Use saturating addition here to avoid wrapping
+		if (~(uint64_t)0 - delay < leeway)
+			delayWithLeeway = ~(uint64_t)0;
+		else
+			delayWithLeeway = delay + leeway;
+
+		if (delayWithLeeway < minDelayWithLeeway) {
+			nextClock = clock;
+			nextDelay = delay;
+			minDelayWithLeeway = delayWithLeeway;
+		}
+	}
+
+	// If we fired a timer, we mustn't wait; the timer code might need to
+	// run in order to set up another timer.
+	if (didFireTimer) {
+#if DEBUG_TIMERS
+		printf("Timer fired, so not waiting\n");
+#endif
+		return 0;
+	}
+
+	// If we aren't waiting for a timer, we want to wait forever on the
+	// completion port.
+	if (nextClock == -1) {
+#if DEBUG_TIMERS
+		printf("Not waiting\n");
+#endif
+		return INFINITE;
+	}
+
+#if DEBUG_TIMERS
+	printf("Waiting for clock %d\n", nextClock);
+#endif
+
+	// Calculate the number of milliseconds we should wait in an ideal world.
+	// Windows can only actually wait multiples of its current tick length,
+	// which defaults to 1/64s, but may vary depending on other programs
+	// that are executing.
+	uint32_t msToWait = nextDelay / 1000000;
+
+#if DEBUG_TIMERS
+	printf("msToWait = %"PRIu32"\n", msToWait);
+#endif
+
+	// Check if we have at least one tick (15ms) of leeway remaining;
+	// if so, use it up.  This is an approximation, but prevents us from
+	// overrunning because of multiple passes through the loop.
+	if (_dispatch_windows_timeout[nextClock].leeway >= 15000000) {
+#if DEBUG_TIMERS
+		printf("In leeway\n");
+#endif
+		if (shouldWait)
+			_dispatch_windows_timeout[nextClock].leeway -= 15000000;
```

Comment:
I don't believe we need to do this any more, since we no longer take multiple passes around the inner loop.